data/TaskDAO.js

```diff
@@ -0,0 +1,70 @@
+const Task = require('./Task');
+// In-memory implementation
+class TaskDAO {
+    constructor() {
+        this.tasks = [];
+        this.maxID = 0;
+    }
+
+    find(criteria) {
+        if(!criteria || Object.keys(criteria).length === 0) return this.tasks;
+        else {
+            return this.tasks.filter(task => this.matchesCriteria(task, criteria));
+        }
+    }
+
+    findById(id) {
+        return this.tasks.filter(task => task.id == id);
+    }
+
+    newTask(props) {
+        if(!this.validateNewTaskProps(props)) return;
+        let taskID = ++this.maxID;
+        let task = new Task(taskID, props.title, props.course, props.deadline, props.completed, props.scheduledDate);
+        this.tasks.push(task);
+        return task;
+    }
+
+    updateTask(props) {
+        if(!this.validateUpdateTaskProps(props)) return;
+        let task = this.findById(props.id)[0];
+        if(task) {
+            task.title = props.title || task.title;
+            task.course = props.course || task.course;
+            task.deadline = props.deadline || task.deadline;
+            task.completed = props.completed || task.completed;
+            task.scheduledDate = props.scheduledDate || task.scheduledDate; 
+        }
+        return task;
+    }
+
+    // might be better to just have a deleted flag?
+    deleteByID(id) {
```

Comment:
FYI - this doesn't delete the item from the array it just fetches and then returns it. Is this done?